File: 2024/March/3-12 1171 Remove Zero Sum Consecutive Nodes From Linked List/RemoveZeroSum.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        /* Base Case */
        if(head == null) {
            return head;
        }

        /* Remove all eligible sublists and connect to head */
        head.next = removeZeroSumSublists(head.next);

        ListNode to = head;
        int sum = 0;

        /* Check for eligibility from head */
        while(to != null) {
            sum += to.val;
            if(sum == 0) {
                return to.next;
            }
            else {
                to = to.next;
            }
        }

        return head;
    }
}

File: 2024/March/3-12 1171 Remove Zero Sum Consecutive Nodes From Linked List/RemoveZeroSum.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var removeZeroSumSublists = function(head) {
    if(!head) {
        return head;
    }

    head.next = removeZeroSumSublists(head.next);

    let to = head;
    let sum = 0;

    while(to) {
        sum += to.val;
        if(sum == 0) {
            return to.next;
        }
        else {
            to = to.next;
        }
    }

    return head;
};